analyzers/src/SonarAnalyzer.Common/SymbolicExecution/Roslyn/OperationProcessors/Invocation.cs

```diff
@@ -171,6 +172,35 @@ private ProgramState ProcessDebugAssertBoolSymbol(ProgramState state, IOperation
         }
     }
 
+    private static ProgramState[] ProcessEquals(SymbolicContext context, IInvocationOperationWrapper invocation)
+    {
+        if (invocation.TargetMethod.IsStatic && invocation.Arguments.Length == 2
+            && invocation.Arguments[0].ToArgument().Value is var leftOperation
+            && invocation.Arguments[1].ToArgument().Value is var rightOperation
+            && context.State[leftOperation]?.Constraint<ObjectConstraint>() is var leftConstraint
+            && context.State[rightOperation]?.Constraint<ObjectConstraint>() is var rightConstraint
+            && (leftConstraint == ObjectConstraint.Null || rightConstraint == ObjectConstraint.Null))
+        {
+            if (leftConstraint == ObjectConstraint.Null && rightConstraint == ObjectConstraint.Null)
+            {
+                return new[] { context.SetOperationConstraint(BoolConstraint.True) };
+            }
+            else if (leftConstraint is not null && rightConstraint is not null)
+            {
+                return new[] { context.SetOperationConstraint(BoolConstraint.False) };
+            }
+            else if ((leftConstraint == ObjectConstraint.Null ? rightOperation : leftOperation).TrackedSymbol() is { } symbol)
+            {
+                return new[]
+                {
+                    context.SetOperationConstraint(BoolConstraint.True).SetSymbolConstraint(symbol, ObjectConstraint.Null),
+                    context.SetOperationConstraint(BoolConstraint.False).SetSymbolConstraint(symbol, ObjectConstraint.NotNull)
+                };
+            }
```

Comment:
This gets much more complicated with nonnullable value types. Propably later.

Reply:
Yep, this is just a small babystep for very specific scenarios